envoy-control-tests/src/main/kotlin/pl/allegro/tech/servicemesh/envoycontrol/permissions/TlsBasedAuthenticationTest.kt

```diff
@@ -51,41 +60,113 @@ internal class TlsBasedAuthenticationTest : EnvoyControlTestConfiguration() {
                     endpoints:
                       - path: "/secured_endpoint"
                         clients: ["echo"]
+                  outgoing:
+                    dependencies:
+                      - service: "echo3"
         """.trimIndent())
 
-        @JvmStatic
-        @BeforeAll
-        fun setupTest() {
-            setup(appFactoryForEc1 = { consulPort ->
-                EnvoyControlRunnerTestApp(properties = properties, consulPort = consulPort)
-            }, envoyConfig = echo1EnvoyConfig, secondEnvoyConfig = echo2EnvoyConfig, envoys = 2)
-        }
+        // language=yaml
+        val envoyContainerWithEcho3SanConfig = Echo3EnvoyAuthConfig.copy(configOverride = """
+            node:
+              metadata:
+                proxy_settings:
+                  outgoing:
+                    dependencies:
+                      - service: "echo2"
+        """.trimIndent())
+
+        // language=yaml
+        private val echo3EnvoyConfigWildcard = Echo3EnvoyAuthConfig.copy(configOverride = """
+            node:
+              metadata:
+                proxy_settings:
+                  incoming:
+                    endpoints:
+                      - path: "/secured_endpoint"
+                        clients: ["*"]
+        """.trimIndent())
+
+        private val envoyNotTrustingCa = Echo1EnvoyAuthConfig.copy(
+                // do not trust default CA used by other Envoys
+                trustedCa = "/app/root-ca2.crt"
+        )
```

Comment:
I'll check but I think that these containers only live during the lifecycle of this class, so creating them to live during one test vs the whole class is not a big difference.

Reply:
The containers live as long as the test class lives so I do not think that's a big problem.

Reply:
The problem as I see it is now we have 9 container launched at the same time and before we had only 3 containers launched at the same time.
It will probably work, but is this really the new, better approach to tests?

Reply:
I think that the problem was not the number of containers running simultaneously but the coupling and spaghetti in `EnvoyControlTestConfiguration` - I can make the short lived containers start and stop at the beginning and end of the test but I do not see why it's better than starting and stopping them at the beginning of the whole class. If someone forgets to stop the container at the end of the test it will not be stopped automatically.

Reply:
@pbetkier can you comment here?

Reply:
> we had only 4 containers launched at the same time.

That is not the case, we had 6 consuls (3 per DC), 2 additional consul agents, and everything else defined in `EnvoyControlTestConfiguration`. so It's more like ~12 vs 9

Reply:
> That is not the case, we had 6 consuls (3 per DC), 2 additional consul agents, and everything else defined in EnvoyControlTestConfiguration. so It's more like ~12 vs 9
I didn't include consul containers in both cases. If you include it you should also include it on the new approach side.
And the point is more general - what if I would like to add another test method with another container. Should I define 10th container in the test setup? I would like to discuss a better approach for it.

> but the coupling and spaghetti in EnvoyControlTestConfiguration
I am not denying that. But I think when we make tests better in some aspects, we should try to not make it worse in other aspects.

>  I do not see why it's better than starting and stopping them at the beginning of the whole class
1. Because more containers running simultaneously consume more resources and there is a limit for that.
2. Because we clutter the test with a setup that is only used in one test method. It makes it less readable.

But as you just show me, there is a simple way to create a container per test method in the new approach. It's all I really wanted.

Reply:
I agree with Marcin's points but I also see that he's already approved changes. Do we agree on keeping clutter? What's this `simple way to create a container per test method in the new approach`?

Reply:
We talked off-github and Marcin's main concern was that he wants to keep the ability to start a container per test, this is possible in this approach and I've done it in one test case. We also talked about implementing `Closeable` so that we do not have to stop containers manually - I'll try it the next time when I do the conversion.